src/components/annual-report/report-data/section4/useFetchSectionFourData.js

```diff
@@ -0,0 +1,111 @@
+    const dataBySubPeriod = await engine.query(section4ByLevel, {
+        variables,
+    })
+    return dataBySubPeriod
+}
+
+export const useFetchSectionFourData = () => {
+    const [loading, setLoading] = useState(false)
+    const [data, setData] = useState(null)
+    const [error, setError] = useState(null)
+
+    const engine = useDataEngine()
+
+    const refetch = async ({ variables = {} }) => {
+        const { denominatorRelations } = variables?.mappedConfiguration
+        const denominatorRelationDEs = [
+            ...denominatorRelations?.map((rel) => rel.A.id),
+            ...denominatorRelations?.map((rel) => rel.B.id),
+        ]
+
+        // set to loading
+        setLoading(true)
+        // get data sets and check that each data element is associated with only one period type
+
+        // we need to preserve information on index to tie response back to relevant denominator relation (some can be filtered out)
+        const byLevelRequestsWithIndices = denominatorRelations
+            .map((rel, index) => {
+                // UN type does not require details by subOrgUnit level, so these can be skipped
+                if (rel.type === 'un') {
+                    return
+                }
+
+                // the level will be the minimum of the lowest of the denominator levels and the selected level
+                const denomMinLevel = Math.min(
+                    rel.A.lowLevel,
+                    rel.B.lowLevel,
+                    variables.orgUnitLevelNumber
+                )
+                // if the denominator relation level is less than the boundary org unit's level, it cannot be calculated
+                if (denomMinLevel < variables.boundaryOrgUnitLevel) {
+                    return
+                }
+                return {
+                    request: fetchDataByLevel({
+                        engine,
+                        variables: {
+                            dataElements: [rel.A.id, rel.B.id],
+                            orgUnits: variables.orgUnits,
+                            orgUnitLevel: `LEVEL-${denomMinLevel}`,
+                            periods: [variables?.currentPeriod.id],
+                        },
+                    }),
+                    relation: { index, ...rel, denomMinLevel },
+                }
+            })
+            .filter((req) => req) //filter out null requests
+        const byLevelRequests = byLevelRequestsWithIndices.map((r) => r.request)
+        const byLevelRequestDenomRelations = byLevelRequestsWithIndices.map(
+            (r) => r.relation
+        )
+        try {
+            const dataByLevel = await Promise.all(byLevelRequests)
+            const overallData = await engine.query(section4Overall, {
+                variables: {
+                    dataElements: denominatorRelationDEs,
+                    orgUnits: variables.orgUnits,
+                    periods: [variables.currentPeriod.id],
+                },
+            })
```

Comment:
Here's a minor optimization that could allow the requests to happen in parallel:

```suggestion
            const overallRequest = engine.query(section4Overall, {
                variables: {
                    dataElements: denominatorRelationDEs,
                    orgUnits: variables.orgUnits,
                    periods: [variables.currentPeriod.id],
                },
            })
            const [overallData, ...dataByLevel] = await Promise.all([
                overallRequest,
                byLevelRequests,
            ])
```